File: bindings/cpp/src/client_feed.cpp
#include "argv.hpp"
#include "kevy/client.hpp"
#include "kevy/errors.hpp"
#include "reply_util.hpp"

// Change feed / CDC: FEED.* (contract §3.10). Both backends serve the same
// cursor contract. This port's embedded connect does not enable the change
// feed (the §5.1 C ABI has no feed-enable flag), so embedded feed_tail/read
// answer Unsupported; feed_shards is always 1 on embedded.

namespace kevy {

using detail::Args;
using detail::raise_reply_error;
using detail::raise_unexpected;
using detail::u2s;

namespace {

FeedFrame parse_feed_frame(Reply& f) {
  if (f.kind != ReplyKind::Array || f.array.size() != 2)
    throw ProtocolError("FEED.READ: frame shape != [offset, argv]");
  Reply& off = f.array[0];
  Reply& argv_raw = f.array[1];
  if (off.kind != ReplyKind::Int || argv_raw.kind != ReplyKind::Array)
    throw ProtocolError("FEED.READ: frame shape != [offset, argv]");
  FeedFrame frame;
  frame.offset = static_cast<uint64_t>(off.integer);
  for (auto& a : argv_raw.array) {
    if (a.kind == ReplyKind::Bulk || a.kind == ReplyKind::Simple) frame.argv.push_back(std::move(a.bytes));
    else raise_unexpected(a);
  }
  return frame;
}

FeedBatch parse_feed_batch(Reply&& r) {
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  if (r.kind != ReplyKind::Array || r.array.size() != 3)
    throw ProtocolError("FEED.READ: expected [gen, next, frames]");
  Reply& g = r.array[0];
  Reply& next = r.array[1];
  Reply& frames = r.array[2];
  if (g.kind != ReplyKind::Int || next.kind != ReplyKind::Int)
    throw ProtocolError("FEED.READ: non-integer cursor");
  if (frames.kind != ReplyKind::Array) throw ProtocolError("FEED.READ: frames not an array");
  FeedBatch batch;
  batch.generation = static_cast<uint64_t>(g.integer);
  batch.next_offset = static_cast<uint64_t>(next.integer);
  for (auto& f : frames.array) batch.frames.push_back(parse_feed_frame(f));
  return batch;
}

}  // namespace

int64_t Client::feed_shards() {
  if (emb_) return 1;
  return exec_count({"FEED.SHARDS"});
}

std::pair<uint64_t, uint64_t> Client::feed_tail(uint64_t shard) {
  if (emb_) {
    if (shard != 0) throw InvalidInputError("embedded feed is single-shard: shard must be 0");
    throw UnsupportedError("feed disabled: this port's embedded connect does not enable the change feed");
  }
  Reply r = exec({"FEED.TAIL", u2s(shard)});
  if (r.kind == ReplyKind::Array && r.array.size() == 2 && r.array[0].kind == ReplyKind::Int &&
      r.array[1].kind == ReplyKind::Int)
    return {static_cast<uint64_t>(r.array[0].integer), static_cast<uint64_t>(r.array[1].integer)};
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  raise_unexpected(r);
}

FeedBatch Client::feed_read(uint64_t shard, uint64_t generation, uint64_t offset,
                            std::optional<uint64_t> count, const ByteList& prefixes) {
  if (emb_) {
    if (shard != 0) throw InvalidInputError("embedded feed is single-shard: shard must be 0");
    throw UnsupportedError("feed disabled: this port's embedded connect does not enable the change feed");
  }
  Args argv;
  argv.reserve(prefixes.size() * 2 + 6);
  argv.add("FEED.READ").add_uint(shard).add_uint(generation).add_uint(offset);
  if (count.has_value() && *count > 0) argv.add("COUNT").add_uint(*count);
  for (auto p : prefixes) argv.add("PREFIX").add(p);
  return parse_feed_batch(exec(argv));
}

}  // namespace kevy

File: bindings/go/client_feed.go
package kevy

import "context"

// Change feed / CDC: FEED.* (contract §3.10). Both backends serve the same
// cursor contract. The embedded backend is single-shard (0) and requires
// the store opened with feed enabled; this port's Connect does not enable
// embedded feed, so embedded feed_tail/read answer Unsupported (matching
// the mem:// case), while feed_shards is always 1.

// FeedFrame is one applied effect's argv at a stream offset (contract §4.6).
type FeedFrame struct {
	Offset uint64
	Argv   [][]byte
}

// FeedBatch is a run of frames plus the cursor to resume from (contract
// §4.7). Resume the next FeedRead from (Generation, NextOffset).
type FeedBatch struct {
	Generation uint64
	NextOffset uint64
	Frames     []FeedFrame
}

// FeedShards returns the number of change-feed shards (embedded: 1).
func (c *Client) FeedShards(ctx context.Context) (int64, error) {
	if c.emb != nil {
		return 1, nil
	}
	return c.execCount(ctx, argv1("FEED.SHARDS"))
}

// FeedTail returns the shard's current (generation, next_offset) cursor —
// where a fresh or resumed consumer begins.
func (c *Client) FeedTail(ctx context.Context, shard int) (generation, nextOffset uint64, err error) {
	if c.emb != nil {
		if shard != 0 {
			return 0, 0, errInvalidInput("embedded feed is single-shard: shard must be 0")
		}
		return 0, 0, errUnsupported("feed disabled: this port's embedded Connect does not enable the change feed")
	}
	r, err := c.exec(ctx, [][]byte{bs("FEED.TAIL"), itob(int64(shard))})
	if err != nil {
		return 0, 0, err
	}
	if r.Kind == KindArray && len(r.Array) == 2 && r.Array[0].Kind == KindInt && r.Array[1].Kind == KindInt {
		return uint64(r.Array[0].Int), uint64(r.Array[1].Int), nil
	}
	if r.Kind == KindError {
		return 0, 0, replyErr(r)
	}
	return 0, 0, unexpectedReply(r)
}

// FeedRead delivers up to count frames past the cursor (server default
// 256, hard cap 4096), optionally key-prefix-filtered (fail-open). An
// unservable cursor surfaces as a Protocol error starting "FEEDRESYNC
// <gen> <tail>"; a cursor ahead of the stream as "ERR feed cursor ahead
// of stream". count <= 0 uses the server default.
func (c *Client) FeedRead(ctx context.Context, shard int, generation, offset uint64, count int, prefixes ...[]byte) (FeedBatch, error) {
	if c.emb != nil {
		if shard != 0 {
			return FeedBatch{}, errInvalidInput("embedded feed is single-shard: shard must be 0")
		}
		return FeedBatch{}, errUnsupported("feed disabled: this port's embedded Connect does not enable the change feed")
	}
	argv := [][]byte{bs("FEED.READ"), itob(int64(shard)), utob(generation), utob(offset)}
	if count > 0 {
		argv = append(argv, bs("COUNT"), itob(int64(count)))
	}
	for _, p := range prefixes {
		argv = append(argv, bs("PREFIX"), p)
	}
	r, err := c.exec(ctx, argv)
	if err != nil {
		return FeedBatch{}, err
	}
	return parseFeedBatch(r)
}

func parseFeedBatch(r Reply) (FeedBatch, error) {
	if r.Kind == KindError {
		return FeedBatch{}, replyErr(r)
	}
	if r.Kind != KindArray || len(r.Array) != 3 {
		return FeedBatch{}, errProtocol("FEED.READ: expected [gen, next, frames]")
	}
	g, next, frames := r.Array[0], r.Array[1], r.Array[2]
	if g.Kind != KindInt || next.Kind != KindInt {
		return FeedBatch{}, errProtocol("FEED.READ: non-integer cursor")
	}
	if frames.Kind != KindArray {
		return FeedBatch{}, errProtocol("FEED.READ: frames not an array")
	}
	batch := FeedBatch{Generation: uint64(g.Int), NextOffset: uint64(next.Int)}
	for _, f := range frames.Array {
		frame, err := parseFeedFrame(f)
		if err != nil {
			return FeedBatch{}, err
		}
		batch.Frames = append(batch.Frames, frame)
	}
	return batch, nil
}

func parseFeedFrame(f Reply) (FeedFrame, error) {
	if f.Kind != KindArray || len(f.Array) != 2 {
		return FeedFrame{}, errProtocol("FEED.READ: frame shape != [offset, argv]")
	}
	off, argvRaw := f.Array[0], f.Array[1]
	if off.Kind != KindInt || argvRaw.Kind != KindArray {
		return FeedFrame{}, errProtocol("FEED.READ: frame shape != [offset, argv]")
	}
	argv := make([][]byte, 0, len(argvRaw.Array))
	for _, a := range argvRaw.Array {
		switch a.Kind {
		case KindBulk, KindSimple:
			argv = append(argv, a.Bytes)
		default:
			return FeedFrame{}, unexpectedReply(a)
		}
	}
	return FeedFrame{Offset: uint64(off.Int), Argv: argv}, nil
}

func utob(n uint64) []byte { return itob(int64(n)) }
